packages/tldraw/src/lib/ui/hooks/useFlatten.ts

```diff
@@ -0,0 +1,191 @@
+import {
+	AssetRecordType,
+	Box,
+	Editor,
+	IndexKey,
+	TLImageAsset,
+	TLImageShape,
+	TLShapeId,
+	Vec,
+	createShapeId,
+	isShapeId,
+	transact,
+	useEditor,
+} from '@tldraw/editor'
+import { useCallback } from 'react'
+
+export async function flattenShapesToImages(
+	editor: Editor,
+	ids: TLShapeId[],
+	multipleImagePadding?: number
+) {
+	if (ids.length === 0) return
+
+	// Don't flatten if it's just one image
+	if (ids.length === 1) {
+		const shape = editor.getShape(ids[0])
+		if (!shape) return
+		if (editor.isShapeOfType(shape, 'image')) return
+	}
+
+	const groups: { ids: TLShapeId[]; bounds: Box; asset?: TLImageAsset }[] = []
+
+	if (multipleImagePadding !== undefined) {
+		const expandedBounds = ids.map((id) => {
+			return {
+				id,
+				bounds: editor.getShapeMaskedPageBounds(id)!.clone().expandBy(multipleImagePadding),
+			}
+		})
+
+		for (let i = 0; i < expandedBounds.length; i++) {
+			const item = expandedBounds[i]
+			if (i === 0) {
+				groups[0] = {
+					ids: [item.id],
+					bounds: item.bounds,
+				}
+				continue
+			}
+
+			let didLand = false
+
+			for (const group of groups) {
+				if (group.bounds.includes(item.bounds)) {
```

Comment:
i'm wondering if we should skip over some shapes. if you flatten an embed right now, it's prbly not what you'd expect. makes me wonder if there should be exceptions, like embeds, or anything else.

Reply:
Good call